### PR TITLE
Improved partition to datastreams assignment:

### PR DESCRIPTION
**Pull request by and-costea**
- take into account when the number of datastreams / node was unbalanced, by having a target # of partitions / node (instead of overall)
- make sure that any assignment is not heavily skewed: the difference between the maximum number of partitions and the minimum should be smaller than a fixed threshold

### CODE REVIEW

File: src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala [hunk 1]
```diff
@@ -56,6 +57,15 @@ class DataStreamPartitionAssignmentTest extends FunSuite with Matchers {
     /** make partitions have affinity to nodes that are not necessarily VectorEndpoints */
     edges <- Gen.listOfN(nA, Gen.resize(replFactor, Gen.listOf(Gen.choose(0, nB * 2 - 1).map(b => s"host-$b"))))
   } yield (edges, generateEndPoints(nB))
+  
+  private val unevenEndpointsPerHostGen = for {
+    nB <- Gen.choose(3, 20)
+    nA <- Gen.choose(0, 1200)
+    /** all partitions are replicated on all nodes */
+    edges = (0 until nA).map(a => (0 until nB).map(b => s"host-$b"))
```
Comment (cbarca) [inline src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala:64]: what is `a`'s role in here? 

just out of curiosity, is it me or you're creating nA x "host-$b" foreach $b in nB strings here? how you're identifying the left nodes from the bigraph then? (is it by the index of the seq itself?)

File: src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala [hunk 2]
```diff
@@ -89,7 +100,20 @@ class DataStreamPartitionAssignmentTest extends FunSuite with Matchers {
       forAllNoShrink(randomMatchParamsGen) {
         case (edges, endpoints) =>
           val assignment = DataStreamPartitionAssignment(edges.toArray, endpoints)
-          assignment.map(_.size).max <= edges.size
+          val ret = assignment.map(_.size).max - assignment.map(_.size).min <= DataStreamPartitionAssignment.MaxSkew
+          if (!ret)
+            logError(s"Assignment is ${assignment.map(_.size)} while number partitions is ${edges.size} and num endpoints is ${endpoints.size}")
```
Comment (cbarca) [inline src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala]: is it enough to log an error here? do you think we need some assertions too about `ret`?

**Comment by cbarca** [comment]
nothing else to add

ship it

File: src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala (hunk 1 above)
Comment (and-costea) [inline src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala:64]: yes, here we just make a complete graph.

File: src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala (hunk 2 above)
Comment (and-costea) [inline src/test/scala/com/actian/spark_vector/datastream/DataStreamPartitionAssignmentTest.scala]: This was just used for debugging. I will remove the logError stuff.

**Pull request merged**